core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -388,6 +388,16 @@ public void updateRoutingEntry(final ShardRouting newRouting, final boolean pers
                     }
                 }
             }
+
+            if (state == IndexShardState.RELOCATED && newRouting.relocating() == false) {
```

Comment:
this is tricky - it leaves us in a potentially scenario where there are two active primaries - the target and the source. I don't have a clean solution for this. My suggestion is to fail the shard and let the master promote another replica.